core/commonMain/src/implementations/immutableSet/PersistentHashSetBuilder.kt

```diff
@@ -43,6 +43,16 @@ internal class PersistentHashSetBuilder<E>(private var set: PersistentHashSet<E>
         return size != this.size
     }
 
+    override fun addAll(elements: Collection<E>): Boolean {
+        if(elements is PersistentHashSet) {
+            val oldSize = this.size
+            size += elements.size
```

Comment:
I don't like it, too, actually, it is unnatural even when you don't consider `modCount`.
But calling `calculateSize()` is supposed to be expensive (it's linear in the size of the node) and my variant only calls it in marginal cases, while yours does it a lot.
It needs to be benchmarked.

Reply:
I've actually put a lot of thought into it precisely with the goal of avoiding calling `calculateSize()` unnecessarily as it seems kinda big deal to me and the `size--` scheme, while not good by any means, is the only thing I came up with.
It is certainly possible to devise an ideal solution here (see the `DeltaCounter` approach in the non-mutable variant) that somehow takes sizes of the roots into account, but it needs more thought.

Reply:
May be a decent tradeoff would be to carry both an intersection counter and the mutator, overwriting the actual size of the mutator in the end. Just a thought.

Reply:
To reproduce `ConcurrentModificationException`:
```
val builder = persistentHashSetOf(1, 2).builder()
val iterator = builder.iterator()
builder.addAll(persistentHashSetOf(1))
iterator.next()  <---
```
No elements were added and internal structure also haven't been changed.

Reply:
Yes, but this is a very questionable example. One could argue that CME should be thrown in this case because the code is wrong by structure. It is very fragile and should not be written like this.